### ISSUE TITLE
Storage.GetClient needs further refactor

### ISSUE THREAD

**Issue by zhiqiangxu** [pingcap/tidb#11773]
## Feature Request

**Is your feature request related to a problem? Please describe:**

Related to add new kv engine to tidb.

**Describe the feature you'd like:**

All other methods in kv.Storage are very clear, except `GetClient()`.It's hard to figure out how to implement this method for another kv engine. So I think it needs further refactor to make it more explicit, so that other kv engines can be integrated with ease.

**Comment by siddontang** [comment]
PTAL @disksing @breeswish

**Comment by disksing** [comment]
Good point. I think we should remove `GetClient` from the `Storage` interface.

**Comment by siddontang** [comment]
thanks @zhiqiangxu 

can you help us do the refactor? if you have any problem, you can ask us for help.

**Comment by zhiqiangxu** [comment]
To do the refactor needs to be well understanding how `CopClient` works, I haven't dig into it yet..

**Comment by disksing** [comment]
I think there is a quick and simple way to fix it, without the need of having in-depth understanding of coprocessor:
- remove `GetClient` in the `Storage` interface directly.
- add new interface `type CopStorage interface { GetClient() Client }`
- for every place that uses `Storage.GetClient`, use type assertion:
```
if cops, ok := s.(CopStorage); ok {
    client := cops.GetClient()
    ...
}
```

**Comment by zhiqiangxu** [comment]
@disksing in the end we still need to implement `CopStorage` for each kv engine, say it's not enough to implement `Storage` interface for an new engine to integrate into tidb.

**Comment by disksing** [comment]
When the engine does not support `CopStorage`, you can turn to use the kv commands instead. We have this mechanism.
In fact, there is another way to resolve your problem. You can implement `GetClient` for the new engine and return a dummy client whose `IsRequestTypeSupported` always returns false. Then SQL layer can automatically use kv commands.

**Comment by zhiqiangxu** [comment]
@disksing Nice to know ! I'll come back when time permits !
